score.go

```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
 	// Use standard parameter defaults (used in Lucene and academic papers)
 	k, b := 1.2, 0.75
 	sumTf := 0.0 // Just for debugging
-	score := 0.0
-	for _, freq := range termFreqs {
+
+	tfs := make(termFrequencyScore)
+
+	for term, freq := range termFreqs {
 		tf := float64(freq)
 		sumTf += tf
-		score += ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
+
+		// Invariant: the keys of df are the union of the keys of tfs over all files.
+		df[term] += 1
+		tfs[term] = ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
 	}
 
-	fileMatch.addBM25Score(score, sumTf, L, opts.DebugScore)
+	if opts.DebugScore {
+		fileMatch.Debug = fmt.Sprintf("sum-termFrequencyScore: %.2f, length-ratio: %.2f", sumTf, L)
+	}
+
+	return tfs
+}
+
+// idf computes the inverse document frequency for a term. nq is the number of
+// documents that contain the term and documentCount is the total number of
+// documents in the corpus.
+func idf(nq, documentCount int) float64 {
+	return math.Log(1.0 + ((float64(documentCount) - float64(nq) + 0.5) / (float64(nq) + 0.5)))
+}
+
+// termDocumentFrequency is a map "term" -> "number of documents that contain the term"
+type termDocumentFrequency map[string]int
+
+// termFrequencyScore is a map "term" -> "term frequency score"
+type termFrequencyScore map[string]float64
+
+// fileMatchesWithScores is a helper type that is used to store the file matches
+// along with internal scoring information.
+type fileMatchesWithScores struct {
```

Comment:
I agree with you that this is a little awkward! If we can't put this on `FileMatch` Could we add the term frequencies to `SearchResult` instead?

Reply:
A pushed a new version in which I add the term frequencies to `FileMatch`. I had to tweak the tests quite to make quick.Check happy. Let me know what you think.

Reply:
Definitely a bit tricky! I don't have a better idea. Tagging @keegancsmith in case he has a strong opinion here.